include/swift/Frontend/CachedDiagnostics.h

```diff
@@ -0,0 +1,56 @@
+//===--- CachedDiagnostics.h - Cached Diagnostics ---------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file defines the CachedDiagnosticConsumer class, which
+//  caches the diagnostics which can be replayed with other DiagnosticConumers.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_CACHEDDIAGNOSTICS_H
+#define SWIFT_CACHEDDIAGNOSTICS_H
+
+#include "llvm/Support/Error.h"
+
+namespace swift {
+
+class CompilerInstance;
+class DiagnosticEngine;
+class SourceManager;
+class FrontendInputsAndOutputs;
+
+class CachingDiagnosticsProcessor {
+public:
+  CachingDiagnosticsProcessor(CompilerInstance &Instance);
+  ~CachingDiagnosticsProcessor();
+
+  /// Start capturing all the diagnostics from DiagnosticsEngine.
+  void startCaptureDiagnostics();
+  /// End capturing all the diagnostics from DiagnosticsEngine.
+  void endCaptureDiagnostics();
+
+  /// Disable catpure.
+  void setDisableCapture();
```

Comment:
I don't understand what this does compared to the above start/end API.  From the implementation I can only disable capture if it's not already started, so what exactly does it mean to disable if it's not capturing?

Reply:
Disable means do not store the output and throw away the capture state. This is for replaying diagnostics.

An empty capture (or never started capture) is no diagnostics and will be stored as no diagnostics in the CAS. While a disable capture doesn't trigger store.

Reply:
Would it make sense for replay to be a `static` method instead? Then you don't need to worry about the capturing "state" in the public API